ethcore/src/miner/service_transaction_checker.rs

```diff
@@ -29,23 +30,24 @@ const SERVICE_TRANSACTION_CONTRACT_REGISTRY_NAME: &'static str = "service_transa
 pub struct ServiceTransactionChecker;
 
 impl ServiceTransactionChecker {
-	/// Checks if given address is whitelisted to send service transactions.
+	/// Checks if given address in tx is whitelisted to send service transactions.
 	pub fn check<C: CallContract + RegistryInfo>(&self, client: &C, tx: &SignedTransaction) -> Result<bool, String> {
 		let sender = tx.sender();
-		let hash = tx.hash();
-
 		// Skip checking the contract if the transaction does not have zero gas price
 		if !tx.gas_price.is_zero() {
 			return Ok(false)
 		}
 
-		let address = client.registry_address(SERVICE_TRANSACTION_CONTRACT_REGISTRY_NAME.to_owned(), BlockId::Latest)
-			.ok_or_else(|| "contract is not configured")?;
-
-		trace!(target: "txqueue", "[{:?}] Checking service transaction checker contract from {}", hash, sender);
+		self.check_address(client, sender)
+	}
 
+	/// Checks if given address is whitelisted to send service transactions.
+	pub fn check_address<C: CallContract + RegistryInfo>(&self, client: &C, sender: Address) -> Result<bool, String> {
+		let contract_address = client.registry_address(SERVICE_TRANSACTION_CONTRACT_REGISTRY_NAME.to_owned(), BlockId::Latest)
```

Comment:
Would the performance be a concern here? `registry_address` would need to query the state if `registrar` is set.

Reply:
I don't think that an issue, some cache would be nice (best would be if previously queried registry entries would be refreshed after a block is imported), but the code is not super hot (only in `\.transact` which is called by the engine, or by the miner, called only for zero-gas-price transactions attempting to enter the pool).